File: backend/app/core/upgrades/stat_modifiers/dash_speed_2.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class DashSpeed2(UpgradeImplementation):
    id = 'dash_speed_2'
    name = 'Swift Escape'
    description = '+32% dash speed'
    rarity = 'epic'
    type = 'stat_modifier'
    cost = 20

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 0.32
        appliers.modifiers.add_multiplier('player', 'dashSpeed', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/dash_speed_2.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const DashSpeed2Def: UpgradeDef = {
  id: "dash_speed_2",
  name: "Swift Escape",
  description: "+32% dash speed",
  rarity: RarityID.Epic,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 20,
  targetClass: UpgradeTargetID.Player,
  fieldInTargetClass: UpgradeStatID.DashSpeed,
  value: 0.32,
  isMultiplier: true,
  stackable: true,
  maxStacks: 2,
  dependentOn: [{ ids: ["dash_ability"] }],
}

export class DashSpeed2 extends Upgrade {}
